lightning/src/ln/msgs.rs

```diff
@@ -79,6 +79,8 @@ pub enum DecodeError {
 	Io(io::ErrorKind),
 	/// The message included zlib-compressed values, which we don't support.
 	UnsupportedCompression,
+	/// The message had invalid semantics.
+	InvalidSemantics(String),
```

Comment:
Hmm... currently this variant is only for logging purposes when failing to parse a message because of invalid semantics. I was thinking `InvoiceError` would only be used after we have a semantically valid message but couldn't handle it for some other reason (e.g., erroring when verifying a message, building an invoice response, or paying an invoice). The spec isn't really clear on this and, to be honest, the TLV portion doesn't seem that well thought out, IMO. My plan was to only use the error message portion.

Also, note that `InvoiceError` is sent back via an onion message to the originator, not to our peer. We'd need to re-think our packet parsing / message handling code to handle this as you're suggesting (i.e., for `DecodeError::InvalidValue`) or at very least only parse the message as TLVs and move semantic checks to the handler (?) in order to reply with an `InvoiceError`. But then we need to expose those TLV types in the handlers where previously they were private.

Reply:
> Why does it need to be a `DecodeError` then? Can't the `OnionMessageHandler` log the `bolt12::ParseError` directly rather than munging it into a `DecodeError` and then logging in `peer_handler`?

Ah, so back in https://github.com/lightningdevkit/rust-lightning/pull/2294#discussion_r1200829954, I think I may have mistaken `Payload` for `Packet` and thus thought the `Logger` would need to be piped through a handful of `ReadableArg` implementations. But `OnionMessenger` handler will directly decode the `Payload` using `onion_utils::decode_next_untagged_hop`. I should be able to pass the logger through there. Let me give that a try.

Reply:
Fixed and rebased. Should be good for another look now.